### PR TITLE
Fix #12 -- support/oppose by region

### LINKED ISSUE #12: Group ballot measure money supporting/opposing by origin (Oakland/California/out of state)

### PR DESCRIPTION
**Pull request by mikeubell**
I have left out the build files that should not be effected by this.  I will add them after a first round of review.

### CODE REVIEW

File: build/referendum/1/supporting/index.json [hunk 1]
```diff
@@ -14,9 +14,18 @@
...
-    "out_of_state": 222
-  }
+  "supporting_locales": [
+    {
+      "amount": 2230.0,
+      "locale": "Within Oakland"
+    },
+    {
+      "amount": 1281607.5,
+      "locale": "Out of State"
```
Comment (tdooner) [inline build/referendum/1/supporting/index.json]: That is some serious money!

File: calculators/referendum_expenditures_by_origin.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      committees.where('"Filer_ID" IS NOT NULL').index_by { |c| c.Filer_ID }
+  end
+
+  def fetch
+    totals = ActiveRecord::Base.connection.execute(<<-SQL)
+      SELECT "Measure_Number", "Sup_Opp_Cd", sum("Amount") AS total
+      FROM "efile_COAK_2016_E-Expenditure",
+        oakland_name_to_number
+      WHERE "Bal_Name" = "Measure_Name"
+      GROUP BY "Measure_Number", "Sup_Opp_Cd";
```
Comment (tdooner) [inline calculators/referendum_expenditures_by_origin.rb]: You need to do the `Report_Num` shenanigans here so we don't get duplicate data

- [ ] add a DISTINCT ON clause for `("Measure_Number", "Sup_Opp_Cd")`
- [ ] add `Report_Num` to the GROUP BY
- [ ] add an ORDER BY `"Measure_Number", "Sup_Opp_Cd", "Report_Num" DESC`

...if I recall correctly

File: calculators/referendum_expenditures_by_origin.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+          SELECT CAST("Filer_ID" as VARCHAR(7)),
+          "Enty_City" as "Tran_City", "Enty_ST" as "Tran_State",
+          "Amount" as "Tran_Amt1"
+          FROM "efile_COAK_2016_497") x
+        WHERE "Filer_ID" IN (SELECT "Filer_ID"
+              FROM "efile_COAK_2016_E-Expenditure"
+              WHERE "Bal_Name" IS NOT NULL)
+        GROUP BY "Filer_ID", locale) A
+      WHERE d."Filer_ID" = A."Filer_ID"
+      ORDER BY d."Measure_Number", d."Sup_Opp_Cd";
```
Comment (tdooner) [inline calculators/referendum_expenditures_by_origin.rb]: Now we're talkin'. I knew we'd get some > 20 line SQL queries in here before too long ;)

File: calculators/referendum_expenditures_by_origin.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+          WHEN lower("Tran_City") = 'oakland' THEN 'Within Oakland'
+          WHEN upper("Tran_State") = 'CA' THEN 'Within California'
+          ELSE 'Out of State'
+        END as locale,
+        sum("Tran_Amt1") as total
+        FROM
+          (SELECT "Filer_ID", "Tran_City", "Tran_State", "Tran_Amt1"
+          FROM "efile_COAK_2016_A-Contributions"
+          UNION
+          SELECT CAST("Filer_ID" as VARCHAR(7)),
```
Comment (tdooner) [inline calculators/referendum_expenditures_by_origin.rb]: `"Filer_ID"::varchar` is a slightly shorter way to cast things in postgres

File: calculators/referendum_expenditures_by_origin.rb (hunk 3 above)
Comment (tdooner) [inline calculators/referendum_expenditures_by_origin.rb]: I hate to say it, but this query will also need to take into account the `Report_Num`.

I'm starting to thing that maybe we should just add a pre-processing step to remove all the old reports.

File: calculators/referendum_expenditures_by_origin.rb [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+          "Enty_City" as "Tran_City", "Enty_ST" as "Tran_State",
+          "Amount" as "Tran_Amt1"
+          FROM "efile_COAK_2016_497") x
+        WHERE "Filer_ID" IN (SELECT "Filer_ID"
+              FROM "efile_COAK_2016_E-Expenditure"
+              WHERE "Bal_Name" IS NOT NULL)
```
Comment (tdooner) [inline calculators/referendum_expenditures_by_origin.rb]: What does this WHERE clause accomplish?

File: calculators/referendum_expenditures_by_origin.rb [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+          totals[measure] -= amount
+          {
+            locale: locale,
+            amount: amount,
+          }
+        end
+        if totals[measure] > 0
+          result <<
+          {
+            locale: 'unknown',
```
Comment (tdooner) [inline calculators/referendum_expenditures_by_origin.rb]: Can you capitalize all the strings in this file, so the frontend can display them without needing to do anything?

Comment (mikeubell) [inline calculators/referendum_expenditures_by_origin.rb]: Will do

File: calculators/referendum_expenditures_by_origin.rb (hunk 2 above)
Comment (mikeubell) [inline calculators/referendum_expenditures_by_origin.rb]: Im not so sure your technique works right.  We can have multiple 460 filings each with their own sequences of Report_Num's.  if we group on Report_Num then each of those groups are only for the 460s that have that Report_Num.  e.g. say there are filings in Jan and June.  The Jan Filing has 1 amendment and the June has 2.  The aggregation will be segmented such that the Report_Num = 2 group will only have the June filing and will not include any data from Jan.

**Pull request merged**